[SPARK-26176][SQL] Verify column names for CTAS with `STORED AS`

sujith71955 opened the pull request:
## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE**
```scala
scala> sql("set spark.sql.hive.convertMetastoreParquet=false")
scala> sql("CREATE TABLE a STORED AS PARQUET AS SELECT 1 AS `COUNT(ID)`")
Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
```

**AFTER**
```scala
scala> sql("CREATE TABLE a STORED AS PARQUET AS SELECT 1 AS `COUNT(ID)`")
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
```

## How was this patch tested? 
Pass the Jenkins with the newly added test case.

sujith71955 committed:
[SPARK-21912][SQL] Invalid column names validation is been added when we create a table using the Hive serde "STORED AS"

 ## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE** `

scala> sql("CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1")

Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
	at java.lang.Enum.valueOf(Enum.java:238)
	at parquet.schema.OriginalType.valueOf(OriginalType.java:21)
	at parquet.schema.MessageTypeParser.addPrimitiveType(MessageTypeParser.java:160)
	at parquet.schema.MessageTypeParser.addType(MessageTypeParser.java:111)
	at parquet.schema.MessageTypeParser.addGroupTypeFields(MessageTypeParser.java:99)
	at parquet.schema.MessageTypeParser.parse(MessageTypeParser.java:92)
	at parquet.schema.MessageTypeParser.parseMessageType(MessageTypeParser.java:82)
	at

**AFTER** ```scala

CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1;

2019-03-13 03:17:58 WARN  ObjectStore:568 - Failed to get database global_temp, returning NoSuchObjectException
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkConversionRequirement(ParquetSchemaConverter.scala:58

## How was this patch tested?  Pass the Jenkins with a new test case.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8813/
Test PASSed.

SparkQA commented:
**[Test build #103390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103390/testReport)** for PR 24075 at commit [`e38e282`](https://github.com/apache/spark/commit/e38e2829ccc98ef1b9924ab452df2e2494d3c355).

SparkQA commented:
**[Test build #103390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103390/testReport)** for PR 24075 at commit [`e38e282`](https://github.com/apache/spark/commit/e38e2829ccc98ef1b9924ab452df2e2494d3c355).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103390/
Test FAILed.

sujith71955 commented:
retest this please

SparkQA commented:
**[Test build #103441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103441/testReport)** for PR 24075 at commit [`6d162fc`](https://github.com/apache/spark/commit/6d162fc190843d56eef3d81698425ef5ce98ddb7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8862/
Test PASSed.

SparkQA commented:
**[Test build #103441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103441/testReport)** for PR 24075 at commit [`6d162fc`](https://github.com/apache/spark/commit/6d162fc190843d56eef3d81698425ef5ce98ddb7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103441/
Test FAILed.

VaibhavDesai commented:
**[Test build Jenkins build has started](https://ci-dev.ngcc.nvidia.com/job/vaibhavd_spark/106/testReport)** for PR 24075 at commit [`d73cfb5`](https://github.com/apache/spark/commit/d73cfb51941f99516b7878acace26db35ea72076).

VaibhavDesai commented:
**[Test build Jenkins build has started](https://ci-dev.ngcc.nvidia.com/job/vaibhavd_spark/106/testReport)** for PR 24075 at commit [`d73cfb5`](https://github.com/apache/spark/commit/d73cfb51941f99516b7878acace26db35ea72076).

sujith71955 commented:
> **[Test build #103390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103390/testReport)** for PR 24075 at commit [`e38e282`](https://github.com/apache/spark/commit/e38e2829ccc98ef1b9924ab452df2e2494d3c355).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

seems to be random failure

sujith71955 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8890/
Test PASSed.

sujith71955 commented:
Above failure is not related to this PR. Please review and let me know for any suugestion. Thanks
cc @HyukjinKwon @dongjoon-hyun @cloud-fan

SparkQA commented:
**[Test build #103473 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103473/testReport)** for PR 24075 at commit [`6d162fc`](https://github.com/apache/spark/commit/6d162fc190843d56eef3d81698425ef5ce98ddb7).

sujith71955 commented:
cc @gatorsmile

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -155,7 +155,7 @@ object HiveAnalysis extends Rule[LogicalPlan] {
 
     case CreateTable(tableDesc, mode, Some(query)) if DDLUtils.isHiveTable(tableDesc) =>
-      DDLUtils.checkDataColNames(tableDesc)
+      DDLUtils.checkDataColNames(tableDesc.copy(schema = query.schema))
```
How is it done for data source tables? by another rule?

sujith71955 replied:
Yes in datasource table scenario, the flow will go through DataSourceAnalysis rule.

sujith71955 replied:
And moreover one more problem what i observed is in serde class name defined "parquet.hive.serde.ParquetHiveSerDe" in checkDataColNames() API,  the serde name shall be "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe" for parquet, so i added this serde also in above code as part of checkDataColNames()  API.

cloud-fan replied:
can we unify this check for both data source table and hive serde table?

sujith71955 replied:
sure, let me check. thanks for your input.

sujith71955 replied:
Both are called from different rules, i will check how to unify

sujith71955 replied:
Handled as part of PreprocessTableCreation rule for CTAS query, please review and let me know for any suggestions. Thanks

dongjoon-hyun commented:
Thank you for pinging me, @sujith71955 . I updated the PR description slightly. I also take a look at this tonight.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8934/
Test PASSed.

SparkQA commented:
**[Test build #103524 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103524/testReport)** for PR 24075 at commit [`6d162fc`](https://github.com/apache/spark/commit/6d162fc190843d56eef3d81698425ef5ce98ddb7).

sujith71955 commented:
> Thank you for pinging me, @sujith71955 .
> 
> * I updated the PR description slightly and triggered a new testing since there was no successful run until now.
> * In addition, I update this JIRA as an `Improvement` since the previous and new behavior are just the same except raising the better exceptions for UX.

Sure. Thanks :)

SparkQA commented:
**[Test build #103524 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103524/testReport)** for PR 24075 at commit [`6d162fc`](https://github.com/apache/spark/commit/6d162fc190843d56eef3d81698425ef5ce98ddb7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103524/
Test PASSed.

sujith71955 committed:
[SPARK-21912][SQL] Verify column names for CTAS with `STORED AS

 ## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE** `

scala> sql("CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1")

Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
	at java.lang.Enum.valueOf(Enum.java:238)
	at parquet.schema.OriginalType.valueOf(OriginalType.java:21)
	at parquet.schema.MessageTypeParser.addPrimitiveType(MessageTypeParser.java:160)
	at parquet.schema.MessageTypeParser.addType(MessageTypeParser.java:111)
	at parquet.schema.MessageTypeParser.addGroupTypeFields(MessageTypeParser.java:99)
	at parquet.schema.MessageTypeParser.parse(MessageTypeParser.java:92)
	at parquet.schema.MessageTypeParser.parseMessageType(MessageTypeParser.java:82)
	at

**AFTER** ```scala

CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1;

2019-03-13 03:17:58 WARN  ObjectStore:568 - Failed to get database global_temp, returning NoSuchObjectException
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkConversionRequirement(ParquetSchemaConverter.scala:58

## How was this patch tested?  Pass the Jenkins with a new test case.

sujith71955 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8977/
Test PASSed.

SparkQA commented:
**[Test build #103568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103568/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).

SparkQA commented:
**[Test build #103568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103568/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103568/
Test FAILed.

sujith71955 commented:
> **[Test build #103568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103568/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

Seems to be failures not related to the PR

sujith71955 commented:
retest this please

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8993/
Test PASSed.

SparkQA commented:
**[Test build #103585 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103585/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).

SparkQA commented:
**[Test build #103585 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103585/testReport)** for PR 24075 at commit [`b199894`](https://github.com/apache/spark/commit/b199894d43277a16f0410ebae9bc7967668008a0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103585/
Test PASSed.

sujith71955 commented:
Gentle ping @dongjoon-hyun @cloud-fan

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala:
```diff
@@ -206,6 +206,8 @@ case class PreprocessTableCreation(sparkSession: SparkSession) extends Rule[Logi
         val analyzedQuery = query.get
         val normalizedTable = normalizeCatalogTable(analyzedQuery.schema, tableDesc)
 
+        DDLUtils.checkDataColNames(tableDesc.copy(schema = analyzedQuery.schema))
```
did we call this in the else branch?

cloud-fan commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -210,7 +209,7 @@ case class RelationConversions(
           if DDLUtils.isHiveTable(tableDesc) && tableDesc.partitionColumnNames.isEmpty &&
             isConvertible(tableDesc) && SQLConf.get.getConf(HiveUtils.CONVERT_METASTORE_CTAS) =>
-        DDLUtils.checkDataColNames(tableDesc)
+        DDLUtils.checkDataColNames(tableDesc.copy(schema = query.schema))
```
Do we need to call it here?

sujith71955 commented:
Nope, i unified this for ctas queries, if non ctas query still flow is
same, means the validation will be done from HiveAnalysis and
DatasourceAnalysis rule. Shall we unify that layer also? Let me know.
Thanks

On Mon, 18 Mar 2019 at 4:33 PM, Wenchen Fan <notifications@github.com>
wrote:

> *@cloud-fan* commented on this pull request.
> ------------------------------
>
> In
> sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala
> <https://github.com/apache/spark/pull/24075#discussion_r266388573>:
>
> > @@ -206,6 +206,8 @@ case class PreprocessTableCreation(sparkSession: SparkSession) extends Rule[Logi
>          val analyzedQuery = query.get
>          val normalizedTable = normalizeCatalogTable(analyzedQuery.schema, tableDesc)
>
> +        DDLUtils.checkDataColNames(tableDesc.copy(schema = analyzedQuery.schema))
>
> did we call this in the else branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/24075#pullrequestreview-215534021>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMZZ-ckORUozuZixUkdWSCUbDBTw9doCks5vX3KWgaJpZM4br_0p>
> .
>

cloud-fan commented:
Yea let's unify that as well

sujith71955 commented:
> Yea let's unify that as well

sure

sujith71955 committed:
[SPARK-21912][SQL] Verify column names for CTAS with `STORED AS

 ## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE** `

scala> sql("CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1")

Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
	at java.lang.Enum.valueOf(Enum.java:238)
	at parquet.schema.OriginalType.valueOf(OriginalType.java:21)
	at parquet.schema.MessageTypeParser.addPrimitiveType(MessageTypeParser.java:160)
	at parquet.schema.MessageTypeParser.addType(MessageTypeParser.java:111)
	at parquet.schema.MessageTypeParser.addGroupTypeFields(MessageTypeParser.java:99)
	at parquet.schema.MessageTypeParser.parse(MessageTypeParser.java:92)
	at parquet.schema.MessageTypeParser.parseMessageType(MessageTypeParser.java:82)
	at

**AFTER** ```scala

CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1;

2019-03-13 03:17:58 WARN  ObjectStore:568 - Failed to get database global_temp, returning NoSuchObjectException
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkConversionRequirement(ParquetSchemaConverter.scala:58

## How was this patch tested?  Pass the Jenkins with a new test case.

sujith71955 commented:
Retest this please.

sujith71955 commented:
@cloud-fan @dongjoon-hyun  Thanks for your valuable time and guidance :)

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9018/
Test PASSed.

SparkQA commented:
**[Test build #103617 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103617/testReport)** for PR 24075 at commit [`02f2c19`](https://github.com/apache/spark/commit/02f2c1985cda4ff806e6778ff54079dd495fa1e4).

SparkQA commented:
**[Test build #103617 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103617/testReport)** for PR 24075 at commit [`02f2c19`](https://github.com/apache/spark/commit/02f2c1985cda4ff806e6778ff54079dd495fa1e4).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103617/
Test FAILed.

dongjoon-hyun commented:
Ur, @sujith71955 
The test failure is relevant to this PR. Could you check once more?
```
org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
```

sujith71955 commented:
> Ur, @sujith71955
> The test failure is relevant to this PR. Could you check once more?
> 
> ```
> org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
> ```

Will check and update the PR. thanks

sujith71955 committed:
[SPARK-21912][SQL] Verify column names for CTAS with `STORED AS

 ## What changes were proposed in this pull request?
Currently, users meet job abortions while creating a table using the Hive serde "STORED AS" with invalid column names. We had better prevent this by raising **AnalysisException** with a guide to use aliases instead like Paquet data source tables.
thus making compatible with error message shown while creating Parquet/ORC native table.

**BEFORE** `

scala> sql("CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1")

Caused by: java.lang.IllegalArgumentException: No enum constant parquet.schema.OriginalType.col1
	at java.lang.Enum.valueOf(Enum.java:238)
	at parquet.schema.OriginalType.valueOf(OriginalType.java:21)
	at parquet.schema.MessageTypeParser.addPrimitiveType(MessageTypeParser.java:160)
	at parquet.schema.MessageTypeParser.addType(MessageTypeParser.java:111)
	at parquet.schema.MessageTypeParser.addGroupTypeFields(MessageTypeParser.java:99)
	at parquet.schema.MessageTypeParser.parse(MessageTypeParser.java:92)
	at parquet.schema.MessageTypeParser.parseMessageType(MessageTypeParser.java:82)
	at

**AFTER** ```scala

CREATE TABLE TAB1TEST STORED AS PARQUET AS SELECT COUNT(ID) FROM TAB1;

2019-03-13 03:17:58 WARN  ObjectStore:568 - Failed to get database global_temp, returning NoSuchObjectException
 Please use alias to rename it.;eption: Attribute name "count(ID)" contains invalid character(s) among " ,;{}()\n\t=".
  at org.apache.spark.sql.execution.datasources.parquet.ParquetSchemaConverter$.checkConversionRequirement(ParquetSchemaConverter.scala:58

## How was this patch tested?  Pass the Jenkins with a new test case.

sujith71955 commented:
> > Ur, @sujith71955
> > The test failure is relevant to this PR. Could you check once more?
> > ```
> > org.apache.spark.sql.hive.execution.SQLQuerySuite.SPARK-21912 ORC/Parquet table should not create invalid column names
> > ```
> 
> Will check and update the PR. thanks

Issue got resolved, we cannot remove the column name validation logic from RelationConversion rule as this rule will run before 'PreprocessTableCreation`, `PreprocessTableInsertion`, `DataSourceAnalysis` and `HiveAnalysis` rules. so before converting the relation we shall do the validation  as per the existing logic.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9059/
Test PASSed.

SparkQA commented:
**[Test build #103660 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103660/testReport)** for PR 24075 at commit [`8da74a9`](https://github.com/apache/spark/commit/8da74a9965128e9541d04ce57d52fabf426c6f9c).

SparkQA commented:
**[Test build #103660 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103660/testReport)** for PR 24075 at commit [`8da74a9`](https://github.com/apache/spark/commit/8da74a9965128e9541d04ce57d52fabf426c6f9c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

sujith71955 commented:
@cloud-fan @dongjoon-hyun  Testcases passed now, let me know for any clarifications/suggestions. thanks

cloud-fan commented:
thanks, merging to master!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103660/
Test PASSed.
